clusterloader2/pkg/measurement/common/network/controller.go

```diff
@@ -0,0 +1,438 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package network
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	clientset "k8s.io/client-go/kubernetes"
+	measurementutil "k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+
+	"k8s.io/klog"
+)
+
+var workerPodList map[string][]workerPodData
+
+var firstClientPodTime int64
+
+const initialDelayForTCExec = 5
+
+var metricVal map[string]MetricResponse
+var uniqPodPairList []uniquePodPair
+var metricRespPendingList []uniquePodPair
+var k8sClient clientset.Interface
+
+var networkPerfRespForDisp NetworkPerfResp
+
+const WorkerListenPort = "5003"
+
+var podPairCh chan uniquePodPair
+
+func (npm *networkPerfMetricsMeasurement) startCtrl() {
+	k8sClient = npm.k8sClient
+	workerPodList = make(map[string][]workerPodData)
+	metricVal = make(map[string]MetricResponse)
+	clientPodNum, _, _ := deriveClientServerPodNum(npm.podRatio)
+	uniqPodPairList = make([]uniquePodPair, 0, clientPodNum)
+	metricRespPendingList = make([]uniquePodPair, 0, clientPodNum)
+	podPairCh = make(chan uniquePodPair)
+}
+
+func populateWorkerPodList(data *workerPodData) {
+	if podData, ok := workerPodList[data.workerNode]; !ok {
+		workerPodList[data.workerNode] = []workerPodData{{podName: data.podName, workerNode: data.workerNode, podIp: data.podIp}}
+	} else {
+		workerPodList[data.workerNode] = append(podData, workerPodData{podName: data.podName, workerNode: data.workerNode, podIp: data.podIp})
+	}
+}
+
+func checkRatioSeparator(ratio string) bool {
+	if !strings.Contains(ratio, RatioSeparator) {
+		return false
+	}
+	return true
+}
+
+func validateCliServPodNum(ratio string) (error, bool) {
+	var podNumber []string
+	var clientPodNum, serverPodNum int
+	podNumber = strings.Split(ratio, RatioSeparator)
+	clientPodNum, _ = strconv.Atoi(podNumber[0])
```

Comment:
should we check the error too?
you are already returning an error if they are negative numbers

Reply:
Yes - don't ignore any errors

Reply:
yes , i have included error handling now.